### PR TITLE
fix null reference

### LINKED ISSUE #1577: Exception while using a UnrelatedTypeEqualityChecks
*@yjbanov commented on May 20, 2019, 3:37 PM UTC:*

Exception from analysis server (running from VSCode / Dart Code)

### What I was doing

Editing code in Flutter's `object.dart` and `box.dart` files, primarily changing type parameters on `RenderObject` and `RenderBox` classes.

### Versions

*   Flutter SDK 1.6.1-pre.46
*   Visual Studio Code 1.34.0
*   Dart Code 3.0.2

### Analyzer Info

The analyzer was launched using the arguments:

```
/flutter/flutter/bin/cache/dart-sdk/bin/snapshots/analysis_server.dart.snapshot
--client-id=Dart-Code.dart-code
--client-version=3.0.2

```

### Exception

Exception while using a UnrelatedTypeEqualityChecks to visit a BinaryExpressionImpl in AssertStatementImpl in BlockImpl in BlockFunctionBodyImpl in MethodDeclarationImpl in ClassDeclarationImpl in CompilationUnitImpl

```
NoSuchMethodError: The getter 'isObject' was called on null.
Receiver: null
Tried calling: isObject
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:50:5)
#1      DartTypeUtilities.unrelatedTypes (package:linter/src/util/dart_type_utilities.dart:370:38)
#2      _hasNonComparableOperands (package:linter/src/rules/unrelated_type_equality_checks.dart:145:25)
#3      _Visitor.visitBinaryExpression (package:linter/src/rules/unrelated_type_equality_checks.dart:191:9)
#4      BinaryExpressionImpl.accept (package:analyzer/src/dart/ast/ast.dart:1003:49)
#5      LinterVisitor._runSubscriptions (package:analyzer/src/lint/linter_visitor.dart:713:14)
#6      LinterVisitor.visitBinaryExpression (package:analyzer/src/lint/linter_visitor.dart:59:5)
#7      BinaryExpressionImpl.accept (package:analyzer/src/dart/ast/ast.dart:1003:49)
#8      AssertStatementImpl.visitChildren (package:analyzer/src/dart/ast/ast.dart:623:17)
#9      RecursiveAstVisitor.visitAssertStatement (package:analyzer/dart/ast/visitor.dart:639:10)
#10     LinterVisitor.visitAssertStatement (package:analyzer/src/lint/linter_visitor.dart:42:11)
#11     AssertStatementImpl.accept (package:analyzer/src/dart/ast/ast.dart:619:49)
#12     NodeListImpl.accept (package:analyzer/src/dart/ast/ast.dart:7568:20)
#13     BlockImpl.visitChildren (package:analyzer/src/dart/ast/ast.dart:1121:17)
#14     RecursiveAstVisitor.visitBlock (package:analyzer/dart/ast/visitor.dart:663:10)
#15     LinterVisitor.visitBlock (package:analyzer/src/lint/linter_visitor.dart:66:11)
#16     BlockImpl.accept (package:analyzer/src/dart/ast/ast.dart:1117:49)
#17     BlockFunctionBodyImpl.visitChildren (package:analyzer/src/dart/ast/ast.dart:1076:13)
#18     RecursiveAstVisitor.visitBlockFunctionBody (package:analyzer/dart/ast/visitor.dart:669:10)
#19     LinterVisitor.visitBlockFunctionBody (package:analyzer/src/lint/linter_visitor.dart:72:11)
#20     BlockFunctionBodyImpl.accept (package:analyzer/src/dart/ast/ast.dart:1072:49)
#21     MethodDeclarationImpl.visitChildren (package:analyzer/src/dart/ast/ast.dart:7008:12)
#22     RecursiveAstVisitor.visitMethodDeclaration (package:analyzer/dart/ast/visitor.dart:1047:10)
#23     LinterVisitor.visitMethodDeclaration (package:analyzer/src/lint/linter_visitor.dart:450:11)
#24     MethodDeclarationImpl.accept (package:analyzer/src/dart/ast/ast.dart:6999:49)
#25     NodeListImpl.accept (package:analyzer/src/dart/ast/ast.dart:7568:20)
#26     ClassDeclarationImpl.visitChildren (package:analyzer/src/dart/ast/ast.dart:1590:13)
#27     RecursiveAstVisitor.visitClassDeclaration (package:analyzer/dart/ast/visitor.dart:699:10)
#28     LinterVisitor.visitClassDeclaration (package:analyzer/src/lint/linter_visitor.dart:102:11)
#29     ClassDeclarationImpl.accept (package:analyzer/src/dart/ast/ast.dart:1560:49)
#30     NodeListImpl.accept (package:analyzer/src/dart/ast/ast.dart:7568:20)
#31     CompilationUnitImpl.visitChildren (package:analyzer/src/dart/ast/ast.dart:2167:21)
#32     RecursiveAstVisitor.visitCompilationUnit (package:analyzer/dart/ast/visitor.dart:723:10)
#33     LinterVisitor.visitCompilationUnit (package:analyzer/src/lint/linter_visitor.dart:126:11)
#34     CompilationUnitImpl.accept (package:analyzer/src/dart/ast/ast.dart:2146:49)
#35     LibraryAnalyzer._computeLints (package:analyzer/src/dart/analysis/library_analyzer.dart:338:10)
#36     LibraryAnalyzer.analyzeSync.<anonymous closure> (package:analyzer/src/dart/analysis/library_analyzer.dart:188:11)
#37     _PerformanceTagImpl.makeCurrentWhile (package:analyzer/src/generated/utilities_general.dart:258:15)
#38     LibraryAnalyzer.analyzeSync (package:analyzer/src/dart/analysis/library_analyzer.dart:183:35)
#39     LibraryAnalyzer.analyze.<anonymous closure> (package:analyzer/src/dart/analysis/library_analyzer.dart:108:14)
#40     _PerformanceTagImpl.makeCurrentWhile (package:analyzer/src/generated/utilities_general.dart:258:15)
#41     LibraryAnalyzer.analyze (package:analyzer/src/dart/analysis/library_analyzer.dart:107:43)
#42     AnalysisDriver._computeAnalysisResult.<anonymous closure> (package:analyzer/src/dart/analysis/driver.dart:1450:63)
#43     PerformanceLog.run (package:analyzer/src/dart/analysis/performance_logger.dart:34:15)
#44     AnalysisDriver._computeAnalysisResult (package:analyzer/src/dart/analysis/driver.dart:1427:20)
#45     AnalysisDriver.performWork (package:analyzer/src/dart/analysis/driver.dart:1247:33)
#46     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:77:64)
#47     _rootRunUnary (dart:async/zone.dart:1132:38)
#48     _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#49     _FutureListener.handleValue (dart:async/future_impl.dart:126:18)
#50     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:639:45)
#51     Future._propagateToListeners (dart:async/future_impl.dart:668:32)
#52     Future._addListener.<anonymous closure> (dart:async/future_impl.dart:348:9)
#53     _rootRun (dart:async/zone.dart:1124:13)
#54     _CustomZone.run (dart:async/zone.dart:1021:19)
#55     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
#56     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
#57     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#58     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#59     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:391:30)
#60     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:416:5)
#61     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:171:12)

```

Diagnostics requested after the error occurred are:

```js
[
    {
        "name": "/flutter/flutter/packages/flutter",
        "explicitFileCount": 970,
        "implicitFileCount": 637,
        "workItemQueueLength": 902,
        "cacheEntryExceptions": []
    }
]
```

*This issue was moved by [pq](https://github.com/pq) from [dart-lang/sdk#37014](https://github.com/dart-lang/sdk/issues/37014).*

### PR DESCRIPTION
**Pull request by pq**
Fixes #1577

/cc @bwilkerson

### CODE REVIEW

File: lib/src/util/dart_type_utilities.dart [hunk 1]
```diff
@@ -367,7 +367,7 @@ class DartTypeUtilities {
         // Otherwise, they might be related.
         return false;
       } else {
-        return leftElement.supertype.isObject ||
+        return leftElement.supertype?.isObject == true ||
```
Comment (yjbanov) [inline lib/src/util/dart_type_utilities.dart]: Test?

Comment (pq) [reply lib/src/util/dart_type_utilities.dart]: Still trying to distill a repro.  Maybe @bwilkerson has ideas?

Comment (bwilkerson) [reply lib/src/util/dart_type_utilities.dart]: Sorry, no. As far as I can remember the supertype ought to always be non-null, so I suspect this is really hiding a bug in analyzer.

Comment (yjbanov) [reply lib/src/util/dart_type_utilities.dart]: According to https://pub.dev/documentation/analyzer/latest/dart_element_element/ClassElement/supertype.html:

> Return the superclass of this class, or `null` if either the class represents the class 'Object' or if the class represents a mixin declaration. All other classes will have a non-`null` superclass.

**Pull request merged**